### PR TITLE
Change getAll permission checking.

### PR DESCRIPTION
**Pull request by mattcollier**
I expect this is a pattern that we will be applying elsewhere.  I wonder if we should add a permission check before the query to determine if the user has PERMISSIONS.LDN_INBOX_ACCESS for any resource at all .
If we do not, a user with no permissions at all will be able to cause the API to return/filter the entire collection just to return [].  We need to figure out what proper protections should be put in place in the HTTP for this reason as well.

### CODE REVIEW

File: lib/index.js [hunk 1]
```diff
@@ -206,24 +206,41 @@ api.inboxes.getAll = function(actor, query, fields, options, callback) {
...
+    for(const record of results.getAuthorized) {
+      if(stripMeta) {
+        delete record.meta;
+      }
       if('message' in record) {
         record.inbox = database.decode(record.inbox);
       }
     }
-    callback(null, results.find);
+    callback(null, results.getAuthorized);
```
Comment (mattcollier) [inline lib/index.js]: We have discussed that it's not a good idea to return a full record.  The getAll endpoint has previously been restricted to admin level permissions.  We are allowing the caller to specify fields to this API, do we want to restrict those fields to properties contained in `inbox` in this case?

Comment (dlongley) [inline lib/index.js]: We don't want to do the restriction here, but we can do it at the HTTP level. It's up to the caller to make sure that the fields are appropriately restricted, i.e. they shouldn't let the actor pick any fields they want.

File: lib/index.js [hunk 2]
```diff
@@ -206,24 +206,41 @@ api.inboxes.getAll = function(actor, query, fields, options, callback) {
 
   query = query || {};
   fields = fields || {};
+  // meta is required for permission check, if meta was
```
Comment (dlongley) [inline lib/index.js]: I think you accidentally your sentence

File: lib/index.js [hunk 3]
```diff
@@ -206,24 +206,41 @@ api.inboxes.getAll = function(actor, query, fields, options, callback) {
...
   options = options || {};
   async.auto({
-    // need inbox access capability for all resources to use this
-    checkPermission: callback => brPermission.checkPermission(
-      actor, PERMISSIONS.LDN_INBOX_ACCESS, callback),
-    find: ['checkPermission', (results, callback) => inboxCollection.find(
-      query, fields, options).toArray(callback)]
+    find: callback => inboxCollection.find(query, fields, options)
+      .toArray(callback),
+    // check to make sure the caller is allowed to access the documentType
```
Comment (dlongley) [inline lib/index.js]: `documentType` should be `inbox`?

File: lib/index.js [hunk 4]
```diff
@@ -206,24 +206,41 @@ api.inboxes.getAll = function(actor, query, fields, options, callback) {
...
+    // (Note: fields *must not* have excluded `documentType.id` in this
```
Comment (dlongley) [inline lib/index.js]: `documentType.id` => `inbox.id`

File: lib/index.js [hunk 5]
```diff
@@ -206,24 +206,41 @@ api.inboxes.getAll = function(actor, query, fields, options, callback) {
 
   query = query || {};
   fields = fields || {};
+  // meta is required for permission check, if meta was
+  let stripMeta = false;
+  if(Object.keys(fields).length > 0 && !fields.meta) {
+    stripMeta = true;
+    fields.meta = true;
```
Comment (dlongley) [inline lib/index.js]: If we're doing this with `meta` -- why not also do it with `id` (which we also need for the permission check below).

File: lib/index.js [hunk 6]
```diff
@@ -206,24 +206,41 @@ api.inboxes.getAll = function(actor, query, fields, options, callback) {
...
-      query, fields, options).toArray(callback)]
+    find: callback => inboxCollection.find(query, fields, options)
+      .toArray(callback),
+    // check to make sure the caller is allowed to access the documentType
+    // (Note: fields *must not* have excluded `documentType.id` in this
+    // case or else the look up for the permission check will fail)
+    getAuthorized: ['find', (results, callback) => async.filterSeries(
+      results.find, (record, callback) => brPermission.checkPermission(
+        actor, PERMISSIONS.LDN_INBOX_ACCESS, {
+          resource: record.meta,
```
Comment (dlongley) [inline lib/index.js]: The resource isn't `record.meta` ... I find this confusing. It should be `record.inbox.id`. If we've already pulled out `meta`, we can just use a list: `resource: [record.inbox.id, record.meta.owner]` and don't use `translate`.

File: lib/index.js [hunk 7]
```diff
@@ -206,24 +206,41 @@ api.inboxes.getAll = function(actor, query, fields, options, callback) {
...
+          translate: 'owner',
+          // record.meta does not include an `id` property
+          excludeId: true
```
Comment (dlongley) [inline lib/index.js]: We can remove this comment and `excludeId` if we use a resource list above that includes the owner and remove translate.

**Review by dlongley — CHANGES_REQUESTED** [review]

File: lib/index.js (hunk 5 above)
Comment (mattcollier) [reply lib/index.js]: Added a todo.

**Comment by mattcollier** [comment]
@dlongley ready for another pass.

**Review by dlongley — APPROVED** [review]

**Pull request merged**